### ISSUE TITLE
how to add the classification task follow the keypoint task?

### ISSUE THREAD

**Issue by xiaohangguo** [ultralytics/ultralytics#3935]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello jocher, great work. As we know, the idea of top-down keypoint detection is very accurate.
1. I would like to classify the keypoints by modifying the model structure, e.g. by adding some fully-connected layers or detection headers at the end - the goal is to get the classification results for body language, is this idea feasible?
2. If it is feasible, should my dataset be a combination of keypoints and categorization dataset? If it is possible please give me some examples.
3. I would like to do target detection - keypoint detection - classification, all at once. i don't know how to start

**Comment by glenn-jocher** [comment]
@xiaohangguo hello,

Thanks for your interest in Ultralytics YOLOv8 and for bringing your questions here. 

To address your first question, adding a classification task following the keypoint detection task can indeed be feasible. As you suggested, modifying the model structure by introducing additional layers, such as fully-connected layers, is a standard approach. Your plan to add detection headers at the end could also work. The important consideration here is to ensure the additional layers are well-integrated and fine-tuned to maintain high performance.

Regarding your second question, you are correct. Your dataset would indeed need to be a combination of keypoints and categorization data. Your examples for each object in your dataset should include both keypoint and category labels. Each example's keypoints labels would indicate the specific points to track in the image for pose estimation while the category labels would indicate what posture or body language the keypoints present.

Your final point about doing target detection, keypoint detection, and classification all at once is quite ambitious, but it is achievable! For this, your model structure would need to be hierarchical or sequential in some form, first performing object detection, then keypoint detection on detected objects, and finally classifying the detected object and keypoints.

To start, you might want to modify the existing YOLOv8 model infrastructure. Start with the target detection task since this is built-in to the core functioning of the YOLOv8 model. Once detection is properly functioning, then you can look at integrating keypoint detection and classification. 

Remember, in each step of your modifications, testing your model and understanding the outputs is key to evaluating your model's performance and making necessary adjustments.

Best of luck with your work! Let us know if you have further questions.

**Comment by glenn-jocher** [comment]
Hello @xiaohangguo,

You're absolutely right about the dataset label formats for object detection and keypoint detection. For object detection, we're generally interested in the bounding box coordinates (x, y, h, w) and for keypoint detection, we label the keypoints by their coordinates (x, y), as you've mentioned.

When you move on to the classification task, things are slightly different. Instead of coordinates, classifications are typically represented by labels or integers. For instance, if you're classifying postures like 'standing', 'sitting', etc., each posture will have a unique integer representation. Let's say 'standing' is represented by the number '0', 'sitting' by '1' and so on. 

In your dataset, after the labels for the object detection and keypoints, you would simply put the integer that corresponds to the classification label directly. So if the classification result here is 'stand', you could have the number '0' (or whatever integer you've assigned to represent 'stand') as the classification label in your dataset.

In summary, the data for each image or sample could be ordered as follows:

1. Object detection data: bounding box coordinates (for example `0 x y h w` ...)
2. Keypoint detection data: keypoint coordinates (for example `x y 0 (1 2),…`)
3. Classification data: category labels (for example `0` for 'stand')

I hope this clears things up! Don't hesitate to reach out if you have more questions. And kudos to your dedication on pushing through your work into the early morning hours!

**Comment by xiaohangguo** [comment]
Thank you very much for your letter. :) I've been following yolo's work for a long time and am a fan of yours! This content has helped me a lot. The next step is to modify the model structure and data reading. 

Unfortunately, in my attempt, an error was reported, it seems to be about the number of channels.
`Traceback (most recent call last):
  File "/home/hang/桌面/yolov8-main/train.py", line 86, in <module>
    model.profile(opt.imgsz)
  File "/home/hang/桌面/yolov8-main/ultralytics/yolo/engine/model.py", line 476, in profile
    return self.model.predict(inputs, profile=True)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/tasks.py", line 67, in predict
    return self._predict_once(x, profile, visualize)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/tasks.py", line 86, in _predict_once
    self._profile_one_layer(m, x, dt)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/tasks.py", line 132, in _profile_one_layer
    m(x.copy() if c else x)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/modules/head.py", line 114, in forward
    kpt = torch.cat([self.cv4[i](x[i]).view(bs, self.nk, -1) for i in range(self.nl)], -1)  # (bs, 17*3, h*w)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/modules/head.py", line 114, in <listcomp>
    kpt = torch.cat([self.cv4[i](x[i]).view(bs, self.nk, -1) for i in range(self.nl)], -1)  # (bs, 17*3, h*w)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/container.py", line 139, in forward
    input = module(input)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/hang/桌面/yolov8-main/ultralytics/nn/modules/conv.py", line 38, in forward
    return self.act(self.bn(self.conv(x)))
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 457, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 453, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,`

`RuntimeError: Given groups=1, weight of size [51, 64, 3, 3], expected input[2, 65, 80, 80] to have 64 channels, but got 65 channels instead`

**Comment by glenn-jocher** [comment]
@xiaohangguo hello,

The error you're encountering seems to stem from a mismatch in the number of channels the layer of the model expects versus the number of channels in the input data. The error message `RuntimeError: Given groups=1, weight of size [51, 64, 3, 3], expected input[2, 65, 80, 80] to have 64 channels, but got 65 channels instead` is effectively saying the model was expecting input with 64 channels but got input with 65 channels instead.

This mismatch could be due to several reasons:

1. Incorrect manipulation or modification of the model architecture; please ensure you didn't accidentally change the channel size of a layer. The change in the expected channel size from 64 to 65 might be an unintended consequence of a modification somewhere else in the model, propagated through.

2. Wrong pre-processing of the input image data. Ensure the images you are feeding into the model have the correct number of channels. 

Try to retrace your steps and double check your modifications to identify where this discrepancy is coming from. 

Let us know if you need further assistance!

**Comment by xiaohangguo** [comment]
> @xiaohangguo hello,
> 
> The error you're encountering seems to stem from a mismatch in the number of channels the layer of the model expects versus the number of channels in the input data. The error message `RuntimeError: Given groups=1, weight of size [51, 64, 3, 3], expected input[2, 65, 80, 80] to have 64 channels, but got 65 channels instead` is effectively saying the model was expecting input with 64 channels but got input with 65 channels instead.
> 
> This mismatch could be due to several reasons:
> 
> 1. Incorrect manipulation or modification of the model architecture; please ensure you didn't accidentally change the channel size of a layer. The change in the expected channel size from 64 to 65 might be an unintended consequence of a modification somewhere else in the model, propagated through.
> 2. Wrong pre-processing of the input image data. Ensure the images you are feeding into the model have the correct number of channels.
> 
> Try to retrace your steps and double check your modifications to identify where this discrepancy is coming from.
> 
> Let us know if you need further assistance!

In fact, it's because I made changes to the source code and I introduced my own modules. Now I'm starting with a new ultralytics code. Unfortunately it still reports an error, the error message seems to be loss, I think maybe I still need to change the loss and the data read in. Can you give me some suggestions?
```
                  from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    221568  ultralytics.nn.modules.conv.Conv             [128, 192, 3, 2]              
  8                  -1  1    259200  ultralytics.nn.modules.block.C2f             [192, 192, 1, True]           
  9                  -1  1    442880  ultralytics.nn.modules.conv.Conv             [192, 256, 3, 2]              
 10                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
 11                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 12                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 13             [-1, 8]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 14                  -1  1    289920  ultralytics.nn.modules.block.C2              [448, 192, 1, False]          
 15                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 16             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 17                  -1  1    131840  ultralytics.nn.modules.block.C2              [320, 128, 1, False]          
 18                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 19             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 20                  -1  1     35200  ultralytics.nn.modules.block.C2              [192, 64, 1, False]           
 21                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 22            [-1, 17]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 23                  -1  1    115456  ultralytics.nn.modules.block.C2              [192, 128, 1, False]          
 24                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 25            [-1, 14]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 26                  -1  1    265344  ultralytics.nn.modules.block.C2              [320, 192, 1, False]          
 27                  -1  1    332160  ultralytics.nn.modules.conv.Conv             [192, 192, 3, 2]              
 28            [-1, 11]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 29                  -1  1    476672  ultralytics.nn.modules.block.C2              [448, 256, 1, False]          
 30    [20, 23, 26, 29]  1   1449976  ultralytics.nn.modules.head.Pose             [1, [17, 3], [64, 128, 192, 256]]
 31                  -1  1    331521  ultralytics.nn.modules.head.Classify         [256, 1]                      
YOLOv8n-pose-p6 summary: 335 layers, 5513673 parameters, 5513657 gradients

Transferred 531/531 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir /home/hang/下载/ultralytics/runs/pose/train', view at http://localhost:6006/
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
Downloading https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8n.pt to 'yolov8n.pt'...
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6.23M/6.23M [00:01<00:00, 4.03MB/s]
AMP: checks passed ✅
train: Scanning /home/hang/下载/datasets/coco8-pose/labels/train... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<00:00, 844.14it/s]
train: New cache created: /home/hang/下载/datasets/coco8-pose/labels/train.cache
val: Scanning /home/hang/下载/datasets/coco8-pose/labels/val... 4 images, 0 backgrounds, 0 corrupt: 100%|██████████| 4/4 [00:00<00:00, 1575.92it/s]
val: New cache created: /home/hang/下载/datasets/coco8-pose/labels/val.cache
Plotting labels to /home/hang/下载/ultralytics/runs/pose/train/labels.jpg... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 84 weight(decay=0.0), 98 weight(decay=0.0005), 97 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 4 dataloader workers
Logging results to /home/hang/下载/ultralytics/runs/pose/train
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss  pose_loss  kobj_loss   cls_loss   dfl_loss  Instances       Size
  0%|          | 0/1 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "/home/hang/anaconda3/envs/yolov8/bin/yolo", line 33, in <module>
    sys.exit(load_entry_point('ultralytics', 'console_scripts', 'yolo')())
  File "/home/hang/下载/ultralytics/ultralytics/cfg/__init__.py", line 423, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/home/hang/下载/ultralytics/ultralytics/engine/model.py", line 376, in train
    self.trainer.train()
  File "/home/hang/下载/ultralytics/ultralytics/engine/trainer.py", line 192, in train
    self._do_train(world_size)
  File "/home/hang/下载/ultralytics/ultralytics/engine/trainer.py", line 332, in _do_train
    self.loss, self.loss_items = self.model(batch)
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/hang/下载/ultralytics/ultralytics/nn/tasks.py", line 44, in forward
    return self.loss(x, *args, **kwargs)
  File "/home/hang/下载/ultralytics/ultralytics/nn/tasks.py", line 212, in loss
    self.criterion = self.init_criterion()
  File "/home/hang/下载/ultralytics/ultralytics/nn/tasks.py", line 327, in init_criterion
    return v8PoseLoss(self)
  File "/home/hang/下载/ultralytics/ultralytics/utils/loss.py", line 301, in __init__
    super().__init__(model)
  File "/home/hang/下载/ultralytics/ultralytics/utils/loss.py", line 117, in __init__
    self.stride = m.stride  # model strides
  File "/home/hang/anaconda3/envs/yolov8/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1207, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'Classify' object has no attribute 'stride'
```
As you can see, I've added another layer of categorization after the pose prediction

**Comment by glenn-jocher** [comment]
Hello @xiaohangguo, 

Thank you for sharing more context about your modifications to the YOLOv8 codebase.

Regarding the error you're encountering, it seems like your 'Classify' layer object does not have the attribute 'stride'. The error originates from this line: `self.stride = m.stride`. 

Here are some areas you might want to investigate:

1. Layer Attributes: Check that you've correctly defined and initialized all the necessary attributes (including 'stride') in your custom 'Classify' layer. 'Stride' in this context is typically associated with Convolutional layers, where it refers to the step size taken when moving the filters across the input.

2. Model Component Interaction: Take a look at how the different components of your model interact with each other. If you are feeding the output of your 'Pose' network into your 'Classify' network and expecting to utilize some of the Pose network's properties in the Classify network (like stride), make sure these properties either exist in the Classify network or are passed correctly to it.

While resolving the above, please remember that since you've introduced custom modules into the source code, the process would involve pinpointing the areas your modifications interact with the existing code base and ensuring they do so as expected.

Hope this helps! Your journey of modifying the original YOLOv8 to perform additional tasks seems exciting and we look forward to seeing your progress! Do keep us posted.

**Comment by xiaohangguo** [comment]
i plan to change the model in your future version like "8.1.0" :) waiting ……

**Comment by thanhhuyenpt** [comment]
> i plan to change the model in your future version like "8.1.0" :) waiting ……
Did you solve your issue? I have the same idea like you. Can you show me did you modifying the model structure to classify the keypoints? Thanks in advance!

**Comment by glenn-jocher** [comment]
@thanhhuyenpt hello there,

It's great to hear that you're planning to modify the future version of the model. Changing the model to classify keypoints is quite an interesting approach.

While I can't provide you with code examples, I can give some general advice on how one might approach this. 

You could begin by creating a new layer or adjusting an existing one to generate keypoint detection on the output. The keypoints may then be classified using a suitable classification algorithm according to the specifications of your project. 

Keep in mind that every modification will impact the model's functionality and potentially its performance, so it is important to test each change thoroughly to make sure it performs as expected.

Please proceed with modifying the model structure and let us know if you encounter any challenges. Your feedback is valuable to us and the community. Good luck with your enhancements!

**Comment by glenn-jocher** [comment]
@xiaohangguo hello there,

It's fantastic to hear that you're both interested in modifying the upcoming version of YOLOv8 to suit your specific project needs.

While I understand the eagerness to dive in, please note that changes to the model structure, especially for classifying keypoints, may require significant amendments to the layers in the network. This kind of modification, moving beyond working with bounding boxes to handle keypoints, would necessitate deeper alterations.

As hinted at, we are indeed working on improvements and updates for future versions such as 8.1. Once it's released, it will certainly provide a fresh baseline for your modifications.

We wish you all the best with implementing your custom changes, and look forward to hearing about your progress. Your pioneering work can serve as inspiration for many other developers in our community. Thank you for your patience and continued support of YOLOv8!
